### PR TITLE
refactor(tags): convert tags to a dictionary

### LINKED ISSUE #4597: Cannot mute failed checks based on AWS tags using 4.3.0
### Steps to Reproduce

1. Create tagged AWS resources (tested with EC2 instances, IAM roles)
2. Add tag-based mute list entry matching those resources
3. Run prowler with the mutelist

### Expected behavior

Findings related to the muted checks are reported as muted for the tagged resources.

### Actual Result with Screenshots or Logs

Related findings are not muted, causing prowler to exit with a non-zero exit code. Additionally, no findings are muted for checks in the mutelist containing both specific resources as well as tags. Meaning not only do tag-based mutelist entries not work, they seem to break the entire mutelist entry for the relevant check.

Note: When the mutelist contains an empty array of tags for a check, the resource-based exclusions still work. Adding an empty string value also has no negative impact. But even adding one of the examples (`environment=dev`) breaks the resource-based muting.

### How did you install Prowler?

From pip package (pip install prowler)

### Environment Resource

Workstation/GitHub Actions

### OS used

`ubuntu-latest` (GitHub Actions)/Fedora 40 Workstation

### Prowler version

4.3.0

### Pip version

24.2

### Context

No warnings/errors are logged regarding our mutelist. When reverting back to 4.2.4, all findings are muted as expected using the exact same mutelist.

I could not find any obvious (to me) reason for the tags breaking while quickly looking through the mutelist code. My best guess is that this issue was somehow introduced by #4413 (just looking for changes to the mutelist added with 4.3.0).

### PR DESCRIPTION
**Pull request by MrCloudSec**
### Description

Convert all Prowler tags to a dictionary for a better formatting and handling of the attribute across all the providers.

### Checklist
- Are there new checks included in this PR? Yes / No
    - If so, do we need to update permissions for the provider? Please review this carefully.
- [x] Review if the code is being covered by tests.
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/tutorials/reporting.md [hunk 1]
```diff
@@ -151,7 +151,7 @@ The JSON-OCSF output format implements the [Detection Finding](https://schema.oc
             "group": {
                 "name": "cloudtrail"
             },
-            "labels": [],
+            "labels": {},
```
Comment (jfagoagas) [inline docs/tutorials/reporting.md]: We cannot do this, this is not compatible with the OCSF format. https://schema.ocsf.io/1.2.0/objects/resource_details?extensions=

File: prowler/lib/mutelist/mutelist.py [hunk 2]
```diff
@@ -65,6 +65,7 @@ def is_muted(
         finding_resource: str,
         finding_tags,
     ) -> bool:
+        print(finding_tags)
```
Comment (jfagoagas) [inline prowler/lib/mutelist/mutelist.py]:
```suggestion
```

File: prowler/lib/outputs/ocsf/ocsf.py [hunk 3]
```diff
@@ -98,11 +98,7 @@ def transform(self, findings: List[Finding]) -> None:
                     resources=[
                         ResourceDetails(
                             # TODO: Check labels for other providers
-                            labels=(
-                                finding.resource_tags.split(",")
-                                if finding.resource_tags
-                                else []
-                            ),
+                            labels=finding.resource_tags,
```
Comment (jfagoagas) [inline prowler/lib/outputs/ocsf/ocsf.py]: This needs to be reverted, `labels` must be a String Array.

File: prowler/lib/outputs/utils.py [hunk 4]
```diff
@@ -57,9 +37,9 @@ def unroll_dict(dict: dict):
         if isinstance(value, list):
             value = ", ".join(value)
         if not unrolled_items:
-            unrolled_items = f"{key}: {value}"
+            unrolled_items = f"{key}={value}"
```
Comment (jfagoagas) [inline prowler/lib/outputs/utils.py]: Why this change in formatting?

File: tests/lib/outputs/csv/csv_test.py [hunk 5]
```diff
@@ -113,7 +113,7 @@ def test_csv_write_to_file(self):
...
+        expected_csv = f"AUTH_METHOD;TIMESTAMP;ACCOUNT_UID;ACCOUNT_NAME;ACCOUNT_EMAIL;ACCOUNT_ORGANIZATION_UID;ACCOUNT_ORGANIZATION_NAME;ACCOUNT_TAGS;FINDING_UID;PROVIDER;CHECK_ID;CHECK_TITLE;CHECK_TYPE;STATUS;STATUS_EXTENDED;MUTED;SERVICE_NAME;SUBSERVICE_NAME;SEVERITY;RESOURCE_TYPE;RESOURCE_UID;RESOURCE_NAME;RESOURCE_DETAILS;RESOURCE_TAGS;PARTITION;REGION;DESCRIPTION;RISK;RELATED_URL;REMEDIATION_RECOMMENDATION_TEXT;REMEDIATION_RECOMMENDATION_URL;REMEDIATION_CODE_NATIVEIAC;REMEDIATION_CODE_TERRAFORM;REMEDIATION_CODE_CLI;REMEDIATION_CODE_OTHER;COMPLIANCE;CATEGORIES;DEPENDS_ON;RELATED_TO;NOTES;PROWLER_VERSION\r\nprofile: default;{datetime.now()};123456789012;123456789012;;test-organization-id;test-organization;test-tag=test-value;test-unique-finding;aws;test-check-id;test-check-id;test-type;PASS;;False;test-service;;high;test-resource;;;;;aws;eu-west-1;check description;test-risk;test-url;;;;;;;test-compliance=test-compliance;test-category;test-dependency;test-related-to;test-notes;{prowler_version}\r\n"
```
Comment (jfagoagas) [inline tests/lib/outputs/csv/csv_test.py]: We cannot change the CSV tags format.

**Review by jfagoagas — CHANGES_REQUESTED** [review]
Not sure about the content of this PR. We need to talk about it.

File: prowler/lib/mutelist/mutelist.py [hunk 6]
```diff
@@ -172,7 +173,7 @@ def is_muted_in_check(
                     muted_in_resource = self.is_item_matched(
                         muted_resources, finding_resource
                     )
-                    muted_in_tags = self.is_item_matched(muted_tags, finding_tags)
+                    muted_in_tags = self.is_tag_matched(muted_tags, finding_tags)
```
Comment (MrCloudSec) [inline prowler/lib/mutelist/mutelist.py]:
```suggestion
                    muted_in_tags = self.is_item_matched(muted_tags, finding_tags)
```

File: prowler/lib/mutelist/mutelist.py [hunk 7]
```diff
@@ -240,7 +241,7 @@ def is_excepted(
                 )
 
                 excepted_tags = exceptions.get("Tags", [])
-                is_tag_excepted = self.is_item_matched(excepted_tags, finding_tags)
+                is_tag_excepted = self.is_tag_matched(excepted_tags, finding_tags)
```
Comment (MrCloudSec) [inline prowler/lib/mutelist/mutelist.py]:
```suggestion
                is_tag_excepted = self.is_item_matched(excepted_tags, finding_tags)
```

File: prowler/lib/mutelist/mutelist.py [hunk 8]
```diff
@@ -290,3 +291,35 @@ def is_item_matched(matched_items, finding_items):
...
+                    if item in finding_tags:
+                        is_item_matched = True
+                        break
+                    # TODO: Cover case of "environment=dev|environment=prod"
+            return is_item_matched
+        except Exception as error:
+            logger.error(
+                f"{error.__class__.__name__} -- {error}[{error.__traceback__.tb_lineno}]"
+            )
+            return False
```
Comment (MrCloudSec) [inline prowler/lib/mutelist/mutelist.py]:
```suggestion
```

File: tests/lib/outputs/csv/csv_test.py [hunk 9]
```diff
@@ -93,7 +93,7 @@ def test_output_transform(self):
         assert output_data["REMEDIATION_CODE_TERRAFORM"] == "terraform-code"
         assert output_data["REMEDIATION_CODE_CLI"] == "cli-code"
         assert output_data["REMEDIATION_CODE_OTHER"] == "other-code"
-        assert output_data["COMPLIANCE"] == "compliance_key: compliance_value"
+        assert output_data["COMPLIANCE"] == "compliance_key=compliance_value"
```
Comment (MrCloudSec) [inline tests/lib/outputs/csv/csv_test.py]:
```suggestion
        assert output_data["COMPLIANCE"] == "compliance_key: compliance_value"
```

File: tests/lib/outputs/csv/csv_test.py (hunk 5 above)
Comment (MrCloudSec) [inline tests/lib/outputs/csv/csv_test.py]:
```suggestion
        expected_csv = f"AUTH_METHOD;TIMESTAMP;ACCOUNT_UID;ACCOUNT_NAME;ACCOUNT_EMAIL;ACCOUNT_ORGANIZATION_UID;ACCOUNT_ORGANIZATION_NAME;ACCOUNT_TAGS;FINDING_UID;PROVIDER;CHECK_ID;CHECK_TITLE;CHECK_TYPE;STATUS;STATUS_EXTENDED;MUTED;SERVICE_NAME;SUBSERVICE_NAME;SEVERITY;RESOURCE_TYPE;RESOURCE_UID;RESOURCE_NAME;RESOURCE_DETAILS;RESOURCE_TAGS;PARTITION;REGION;DESCRIPTION;RISK;RELATED_URL;REMEDIATION_RECOMMENDATION_TEXT;REMEDIATION_RECOMMENDATION_URL;REMEDIATION_CODE_NATIVEIAC;REMEDIATION_CODE_TERRAFORM;REMEDIATION_CODE_CLI;REMEDIATION_CODE_OTHER;COMPLIANCE;CATEGORIES;DEPENDS_ON;RELATED_TO;NOTES;PROWLER_VERSION\r\nprofile: default;{datetime.now()};123456789012;123456789012;;test-organization-id;test-organization;test-tag:test-value;test-unique-finding;aws;test-check-id;test-check-id;test-type;PASS;;False;test-service;;high;test-resource;;;;;aws;eu-west-1;check description;test-risk;test-url;;;;;;;test-compliance: test-compliance;test-category;test-dependency;test-related-to;test-notes;{prowler_version}\r\n"
```

File: tests/lib/outputs/csv/csv_test.py [hunk 10]
```diff
@@ -193,7 +193,7 @@ def test_csv_custom_file_descriptor(self):
...
+            expected_csv = f"AUTH_METHOD;TIMESTAMP;ACCOUNT_UID;ACCOUNT_NAME;ACCOUNT_EMAIL;ACCOUNT_ORGANIZATION_UID;ACCOUNT_ORGANIZATION_NAME;ACCOUNT_TAGS;FINDING_UID;PROVIDER;CHECK_ID;CHECK_TITLE;CHECK_TYPE;STATUS;STATUS_EXTENDED;MUTED;SERVICE_NAME;SUBSERVICE_NAME;SEVERITY;RESOURCE_TYPE;RESOURCE_UID;RESOURCE_NAME;RESOURCE_DETAILS;RESOURCE_TAGS;PARTITION;REGION;DESCRIPTION;RISK;RELATED_URL;REMEDIATION_RECOMMENDATION_TEXT;REMEDIATION_RECOMMENDATION_URL;REMEDIATION_CODE_NATIVEIAC;REMEDIATION_CODE_TERRAFORM;REMEDIATION_CODE_CLI;REMEDIATION_CODE_OTHER;COMPLIANCE;CATEGORIES;DEPENDS_ON;RELATED_TO;NOTES;PROWLER_VERSION\nprofile: default;{datetime.now()};123456789012;123456789012;;test-organization-id;test-organization;test-tag=test-value;test-unique-finding;aws;test-check-id;test-check-id;test-type;PASS;;False;test-service;;high;test-resource;;;;;aws;eu-west-1;check description;test-risk;test-url;;;;;;;test-compliance=test-compliance;test-category;test-dependency;test-related-to;test-notes;{prowler_version}\n"
```
Comment (MrCloudSec) [inline tests/lib/outputs/csv/csv_test.py]:
```suggestion
            expected_csv = f"AUTH_METHOD;TIMESTAMP;ACCOUNT_UID;ACCOUNT_NAME;ACCOUNT_EMAIL;ACCOUNT_ORGANIZATION_UID;ACCOUNT_ORGANIZATION_NAME;ACCOUNT_TAGS;FINDING_UID;PROVIDER;CHECK_ID;CHECK_TITLE;CHECK_TYPE;STATUS;STATUS_EXTENDED;MUTED;SERVICE_NAME;SUBSERVICE_NAME;SEVERITY;RESOURCE_TYPE;RESOURCE_UID;RESOURCE_NAME;RESOURCE_DETAILS;RESOURCE_TAGS;PARTITION;REGION;DESCRIPTION;RISK;RELATED_URL;REMEDIATION_RECOMMENDATION_TEXT;REMEDIATION_RECOMMENDATION_URL;REMEDIATION_CODE_NATIVEIAC;REMEDIATION_CODE_TERRAFORM;REMEDIATION_CODE_CLI;REMEDIATION_CODE_OTHER;COMPLIANCE;CATEGORIES;DEPENDS_ON;RELATED_TO;NOTES;PROWLER_VERSION\nprofile: default;{datetime.now()};123456789012;123456789012;;test-organization-id;test-organization;test-tag:test-value;test-unique-finding;aws;test-check-id;test-check-id;test-type;PASS;;False;test-service;;high;test-resource;;;;;aws;eu-west-1;check description;test-risk;test-url;;;;;;;test-compliance: test-compliance;test-category;test-dependency;test-related-to;test-notes;{prowler_version}\n"
```

File: tests/lib/outputs/asff/asff_test.py [hunk 11]
```diff
@@ -70,7 +70,7 @@ def test_asff(self):
                     Type=finding.resource_type,
                     Partition=AWS_COMMERCIAL_PARTITION,
                     Region=AWS_REGION_EU_WEST_1,
-                    Tags=ASFF.format_resource_tags(finding.resource_tags),
+                    Tags=(finding.resource_tags if finding.resource_tags else None),
```
Comment (jfagoagas) [inline tests/lib/outputs/asff/asff_test.py]:
```suggestion
                    Tags=finding.resource_tags,
```

File: tests/lib/outputs/asff/asff_test.py [hunk 12]
```diff
@@ -136,7 +136,7 @@ def test_asff_without_remediation_recommendation_url(self):
...
```
Comment (jfagoagas) [inline tests/lib/outputs/asff/asff_test.py]:
```suggestion
                    Tags=finding.resource_tags,
```

File: tests/lib/outputs/asff/asff_test.py [hunk 13]
```diff
@@ -205,7 +205,7 @@ def test_asff_with_long_description_and_remediation_recommendation_text(
...
```
Comment (jfagoagas) [inline tests/lib/outputs/asff/asff_test.py]:
```suggestion
                    Tags=finding.resource_tags,
```

File: tests/lib/outputs/asff/asff_test.py [hunk 14]
```diff
@@ -412,7 +412,7 @@ def test_fill_json_asff_with_long_associated_standards(self):
...
```
Comment (jfagoagas) [inline tests/lib/outputs/asff/asff_test.py]:
```suggestion
                    Tags=finding.resource_tags,
```

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: prowler/lib/outputs/asff/asff.py [hunk 15]
```diff
@@ -99,7 +96,11 @@ def transform(self, findings: list[Finding]) -> None:
                                 Type=finding.resource_type,
                                 Partition=finding.partition,
                                 Region=finding.region,
-                                Tags=resource_tags,
+                                Tags=(
+                                    finding.resource_tags
+                                    if finding.resource_tags
+                                    else None
+                                ),
```
Comment (MrCloudSec) [inline prowler/lib/outputs/asff/asff.py]:
```suggestion
                                Tags=finding.resource_tags,
```

**Review by jfagoagas — CHANGES_REQUESTED** [review]
Could you please add the finding type to the `is_finding_muted` method in each provider's Mutelist class?

**Review by jfagoagas — DISMISSED** [review]
Now I can approve it, my 🧠 can rest since the tests covers all the changes in tags and everything works 💯 🚀

**Comment by pedrooot** [comment]
🔝 @jfagoagas 
programming

**Review by jfagoagas — APPROVED** [review]

**Pull request merged**